identity/app/controllers/EmailVerificationController.scala

```diff
@@ -24,28 +25,84 @@ class EmailVerificationController(
     with ImplicitControllerExecutionContext
     with SafeLogging {
 
-  import authenticatedActions.fullAuthWithIdapiUserAction
-
   def completeRegistration(): Action[AnyContent] =
-    fullAuthWithIdapiUserAction.async { implicit request =>
-      val idRequest = idRequestParser(request)
+    Action.async { implicit request =>
       val page = IdentityPage("/complete-registration", "Complete Signup", isFlow = true)
       val verifiedReturnUrlAsOpt = returnUrlVerifier.getVerifiedReturnUrl(request)
 
-      Future.successful(
-        Ok(
-          IdentityHtmlPage.html(
-            views.html.verificationEmailResent(
-              request.user,
-              idRequest,
-              idUrlBuilder,
-              verifiedReturnUrlAsOpt,
-              returnUrlVerifier.defaultReturnUrl,
-              isSignupFlow = true,
-            ),
-          )(page, request, context),
-        ),
-      )
+      request
+        .getQueryString("encryptedEmail")
+        .map(encryptedEmail => {
+          api.decryptEmailToken(encryptedEmail) map {
+            case Left(errors) =>
+              logger.error(s"Could not decrypt email address on complete registration page: $errors")
+              errorPage(verifiedReturnUrlAsOpt, page)
+            case Right(email) =>
+              successPage(verifiedReturnUrlAsOpt, page, Some(email))
+                .withSession("encryptedEmail" -> encryptedEmail, "email" -> email)
+          }
+        })
+        .getOrElse({
+          logger.error("No encryptedEmail parameter present on complete registration page request")
+          Future.successful(errorPage(verifiedReturnUrlAsOpt, page))
+        })
+    }
+
+  def resendValidationEmail(): Action[AnyContent] = {
+    Action.async { implicit request =>
+      val page = IdentityPage("/resend-validation", "Resend Validation", isFlow = true)
+      val verifiedReturnUrlAsOpt = returnUrlVerifier.getVerifiedReturnUrl(request)
+      val email = request.session.get("email")
+
+      request.session
+        .get("encryptedEmail")
+        .map(token =>
+          api.resendEmailValidationEmailByToken(token, verifiedReturnUrlAsOpt) map {
+            case Left(errors) =>
+              logger.error(s"Could not resent validation email on complete registration page: $errors")
```

Comment:
it might be worth returning an error page in this scenario

Reply:
This is what the existing page does at present. It wasn't included in the card to change this and I can see arguments for both behaviours, but I guess if there is a temporary error (such as IDAPI being unavailable or a network issue) they then have the opportunity to resend the email which they wouldn't have on the error page